pkg/image/origins/main.go

```diff
@@ -0,0 +1,201 @@
+
+	var unusedImages []string
+	for k := range img.OriginMap {
+		_, ok := currentImages[k]
+		if !ok {
+			unusedImages = append(unusedImages, k)
+		}
+	}
+
+	return unusedImages
+}
+
+func PrintUpdatedImageOrigins(targetsAndSources utilities.ImageTargetsAndSources) error {
+	fmt.Println("Generating updated rancher-image-origins map")
+
+	// use the existing map so that we don't
+	// check images which have already been resolved.
+	imgToURL := img.OriginMap
+
+	// look through any images passed as arguments
+	err := convertImagesToRepoUrls(targetsAndSources.LinuxImagesFromArgs, imgToURL)
+	if err != nil {
+		return err
+	}
+	// look through the linux target images
+	err = convertImagesToRepoUrls(targetsAndSources.TargetLinuxImages, imgToURL)
+	if err != nil {
+		return err
+	}
+	// look through the windows target images
+	err = convertImagesToRepoUrls(targetsAndSources.TargetWindowsImages, imgToURL)
+	if err != nil {
+		return err
+	}
+
+	// order images alphabetically
+	type sorter struct {
+		img string
+		url string
+	}
+	var sorted []sorter
+	for k, v := range imgToURL {
+		sorted = append(sorted, sorter{img: k, url: v})
+	}
+
+	sort.Slice(sorted, func(i, j int) bool {
+		return sorted[i].img < sorted[j].img
+	})
+
+	x := "OriginMap = map[string]string{\n"
+	for _, e := range sorted {
+		x = x + "	\"" + e.img + "\":" + " " + "\"" + e.url + "\"," + "\n"
+	}
+	x = x + "}\n"
+
+	// print result, can be copied into pkg/image/origins.go
+	// to automatically update some images.
+	fmt.Println(x)
+
+	// find all images that have not been resolved yet
+	var unknownImages string
+	for k, v := range imgToURL {
+		if v == "" || v == "unknown" {
+			unknownImages = unknownImages + "\n" + k
+		}
+	}
+
+	// warn about unresolved images
+	// so that they may be manually resolved
+	if len(unknownImages) > 0 {
+		fmt.Println(fmt.Sprintf("[WARN] Some images do not have an origin defined, please provide origins within rancher/pkg/image/origins.go for the following images: %s", unknownImages))
+	}
+
+	return nil
+}
+
+// convertImagesToRepoUrls attempts to resolve Rancher owned images with its origin repository.
+// It does so on a best-effort basis, any gaps left by this must be covered manually
+// in the OriginMap in origins.go
+func convertImagesToRepoUrls(images []string, imgToURL map[string]string) error {
+	for _, repo := range img.UniqueTargetImages(images) {
+		if v, ok := imgToURL[repo]; ok && (v != "" && v != "unknown") {
+			continue // image is already resolved
+		}
+		switch {
+		case strings.Contains(repo, "mirrored"):
+			// mirrored images, assumes everything after 'mirrored-'
+			// can be directly converted into a github repository.
+			ownerAndRepository := strings.Split(strings.ReplaceAll(repo, "mirrored-", ""), "-")
+			if len(ownerAndRepository) <= 1 {
+				imgToURL[repo] = "unknown"
+				continue
+			}
+			url, err := checkURL(fmt.Sprintf("https://www.github.com/%s/%s", ownerAndRepository[0], strings.Join(ownerAndRepository[1:], "")))
```

Comment:
`https://www.github.com` just 301s to `https://github.com`. I don't think it will be a measurable time save, but might be worth trying.